Refactor scan

jgustie opened the pull request:
This PR serves two purposes: most importantly it moves experiment generation out of `redskyctl`. It also makes "scanning" independent of both "experiment generation" and Kustomize. A few Kustomize API references are left, but in large part the pipeline has been converted from using a Kustomize `ResMap` and API Machinery objects over to using KYAML.

The immediate benefit is that the code is significantly less complex (most everything boils down to a KYAML `Filter` function), this will make improvements to experiment generation easier. Output formatting is also improved (e.g. field ordering is more consistent). The use of KYAML pipeline also makes it easier to load resource manifests from a variety of inputs.

jgustie committed:
Move application and experiment generation code to internal/

jgustie committed:
Fix typo

jgustie committed:
Initial pass at application scanning clean up

jgustie committed:
WIP : Isolate what is being scanned

jgustie committed:
Just rewrite the scanner

jgustie committed:
Move the generator up a level

jgustie committed:
Preemptive strike against #280

jgustie committed:
Improve test for StormForger organization

jgustie committed:
Fix bug with test case file, adjust output ordering

jgustie committed:
Allow scan to express both runtime to YAML and back

jgustie committed:
Do not revert YAML field order changes

jgustie committed:
Split up the scan code into multiple source files

jgustie committed:
Fix copyright spacing

bradbeam commented on internal/scan/scan.go:
```diff
@@ -0,0 +1,78 @@
+// turns them back into resource nodes. The original resource nodes used
+// to generate the inputs are also made available.
+type Transformer interface {
+	Filter(nodes []*yaml.RNode, selected []interface{}) ([]*yaml.RNode, error)
```
nit, should this be `Transform` instead of filter?
Thinking about reducing friction between Transformer.Filter and kio.Filter from the scanner implementation, ex
scanner.Transformer.Filter v scanner.Filter

bradbeam commented on internal/scan/runtime.go:
```diff
@@ -0,0 +1,143 @@
+
+func init() {
+	_ = clientgoscheme.AddToScheme(Scheme)
+	_ = redskyv1beta1.AddToScheme(Scheme)
```
I'm assuming we can ignore redsky.v1alpha1 because the contract for the v1alpha1.application is to only generate redsky.v1beta1?

jgustie replied:
Originally I only had `ObjectSlice` using this scheme so it made sense to leave it off (i.e. you wouldn't generate content from the v1alpha1 structs). But, yeah with the `ObjectList` it makes sense to be there.

bradbeam commented on internal/scan/runtime.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+// ObjectSlices allows a slice of object instances to be read as resource nodes.
+type ObjectSlice []runtime.Object
```
should we add in the interface assertions here as well?
```
var (
  _ kio.Reader = ObjectSlice
  _ kio.Writer = ObjectList
  )
```

if it's not worth doing the above, could we add a comment to each of these types to designate the intent?

bradbeam commented on internal/scan/runtime.go:
```diff
@@ -0,0 +1,143 @@
+			return err
+		}
+
+		obj := newObjectForKind(u)
+		if err := Scheme.Convert(u, obj, runtime.InternalGroupVersioner); err != nil {
+			return err
+		}
```
does this cause an error when `obj == u` ( as in newObjectForKind did not find a match )?
trying to understand the expected behavior in this situation.

jgustie replied:
When `obj == u` convert basically just does a shallow copy of the backing map, it's not an error. The end result is the unstructured object is wrapped in a `RawExtension` in the list (an unstructured object _is_ a `runtime.Object`). Code that does type checks on `Items[n].Object` will see `*unstructured.Unstructured` for any kind that does not have a Go type registered in the `Scheme`.

bradbeam commented on internal/scan/selector.go:
```diff
@@ -0,0 +1,151 @@
+
+// compileAnchored is a helper to make regular expressions match a full string.
+func compileAnchored(pattern string) (*regexp.Regexp, error) {
+	if pattern == "" {
+		return nil, nil
+	}
```
instead of returning nil here, should we do a catchall regex ( ex, `*` or `.*` )?
that should allow us to skip the `regex != nil` checks in matchesMeta

jgustie replied:
It just seemed like unnecessary extra use of regular expression matching in a loop. Maybe I'm just dating myself.

bradbeam commented on internal/scan/selector.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// metaMatcher matches metadata. This is an alternative to `types.SelectorRegex` from Kustomize.
+type metaMatcher struct {
+	namespaceRegex *regexp.Regexp
+	nameRegex      *regexp.Regexp
+	groupRegex     *regexp.Regexp
+	versionRegex   *regexp.Regexp
+	kindRegex      *regexp.Regexp
+}
```
why use regex instead of strings here?

as in why wouldn't we just save off `g.Namespace` ( or `strings.Split(g.Namespace, ":")[1])` ) so we could compare `string == string` in `matchesMeta`

at the end of the day I think either implementation would work, just curious.

jgustie replied:
We allow things like `Group: "apps|extensions"`. This behavior was previously provided by Kustomize via its `types.Selector` which is used by the `ResMap` to select subsets of resources.

bradbeam commented on internal/scan/selector.go:
```diff
@@ -0,0 +1,151 @@
+	}
+
+	if m.groupRegex != nil || m.versionRegex != nil {
+		group, version := "", meta.APIVersion
+		if pos := strings.Index(version, "/"); pos >= 0 {
+			group, version = version[0:pos], version[pos+1:]
+		}
```
should we use https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime/schema#ParseGroupVersion instead?
ultimately the same, so indifferent on this suggestion

bradbeam commented on internal/scan/selector.go:
```diff
@@ -0,0 +1,151 @@
+
+	if m.groupRegex != nil || m.versionRegex != nil {
+		group, version := "", meta.APIVersion
+		if pos := strings.Index(version, "/"); pos >= 0 {
+			group, version = version[0:pos], version[pos+1:]
```
should we switch this to `pos > 0` just to clarify if pos == 0 then the values aren't changed from the defaults above?

Or should we just start with this?
```
pos := strings.Index(meta.APIVersion, "/")
group, version = version[0:pos], version[pos+1:]
```

jgustie replied:
They are still changed if `pos == 0`, the leading `/` is removed, e.g. `/v1` should give `"", "v1"` not `"", "/v1"`. This is consistent with the `ParseGroupVersion` function you mention above. I'll add some tests.

bradbeam reviewed (commented):
comments around internal/scan
(still need to check out internal/experiment )

bradbeam commented on internal/experiment/generator.go:
```diff
@@ -0,0 +1,152 @@
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// Generator is use to create an experiment definition.
```
sp `s/use/used`

bradbeam commented on internal/experiment/generator.go:
```diff
@@ -0,0 +1,152 @@
+	g.ContainerResourcesSelectors = []generation.ContainerResourcesSelector{
+		{
+			GenericSelector: scan.GenericSelector{
+				Group:         "apps|extensions",
```
Not sure I caught this in the scan implementation, but why is the format `x|y`?

jgustie replied:
This is the use of the regexp, in this case we are matching two groups from a single selector (in the previous version of the code, I had neglected to leverage the regular expression matching and had just added the selector twice: once for `apps` and once for `extensions`).

bradbeam commented on internal/experiment/generator.go:
```diff
@@ -0,0 +1,152 @@
+			&konjure.Filter{
+				Depth:         100,
+				DefaultReader: g.DefaultReader,
+				KeepStatus:    true,
```
Should we set `KeepStatus: false` if we're going to strip it below?

jgustie replied:
Keeping the status here ensures objects fetched from Kubernetes using `kubectl` retain their status so we can inspect it during the scan. This is independent from striping the status later (which would ensure the experiment itself doesn't have a status field). In the case where `IncludeApplicationResources == true`, then it would also strip the status of the objects fetched from Kube, but only after giving us the opportunity to scan them.

bradbeam commented on internal/experiment/generator.go:
```diff
@@ -0,0 +1,152 @@
+		},
+	}
+
+	for i := range g.ContainerResourcesSelectors {
+		scanner.Selectors = append(scanner.Selectors, &g.ContainerResourcesSelectors[i])
+	}
```
should we skip the range and do 
```
scanner.Selectors = append(scanner.Selectors, &g.ContainerResourcesSelectors...)
```

jgustie replied:
No, because the elements need to have their type converted independently.

bradbeam replied:
Doesn't that get handled by the variadic parameter?

bradbeam replied:
err, I think I got what you're saying.

jgustie replied:
https://golang.org/doc/faq#convert_slice_of_interface

jgustie committed:
Code review comments

jgustie committed:
Explicitly exclude the initialObjectiveCount field

Even though the field is not exported, the path through KYAML JSON serialization picks it up.

jgustie committed:
Must export initialObjectiveCount to suppress it from output

The Kube unstructured converter leverages json: tags but does not ignore un-exported fields. Since vet does not allow json: tags on un-exported fields the easiest solution is to just export the field.

An alternate solution would be to implement json.Marshaller which would bypass the default unstructured conversion code in favor a JSON round trip.

bradbeam commented on internal/experiment/generation/containerresources.go:
```diff
@@ -0,0 +1,190 @@
+
+	containerName, err := regexp.Compile("^(?:" + s.ContainerName + ")$")
+	if err != nil {
+		// Kustomize panics. Not sure where it validates. We will just fall back to exact match
```
Have any more context/deets around this? Not sure I'm following where kustomize would panic from regex.Compile :)

jgustie replied:
This is an old comment I probably should have cleaned up. It's in reference to the fact that Kustomize uses `MustCompile` instead of `Compile`. I never did find the code that did the error checking to give a useful error message though.

bradbeam commented on internal/experiment/generation/generation.go:
```diff
@@ -0,0 +1,102 @@
+
+	redskyappsv1alpha1 "github.com/thestormforge/optimize-controller/api/apps/v1alpha1"
+	redskyv1beta1 "github.com/thestormforge/optimize-controller/api/v1beta1"
+	"github.com/yujunz/go-getter"
```
did you intend this go-getter ( instead of hashicorp )?

jgustie replied:
This is the fork of go-getter that Kustomize (previously) used so it's already in `go.sum`. We can revisit when we bump to the 4.x Kustomize API and it drops out of the transient dependencies.

bradbeam commented on internal/experiment/generation/generation.go:
```diff
@@ -0,0 +1,102 @@
+	yaml.FieldOrder["patch"] = 200
+	yaml.FieldOrder["baseline"] = 100
+	yaml.FieldOrder["min"] = 200
+	yaml.FieldOrder["max"] = 300
```
nit, could we group up the priorities ( 100 / 200 / 300 )

jgustie replied:
I grouped them by attribute set instead of priority, e.g. parameters before metrics. I'll add comments.

jgustie replied:
I've got another idea. I'll just iterate over the type's fields and add the `json:` tag names in the order that they appear...

bradbeam commented on internal/experiment/generation/generation.go:
```diff
@@ -0,0 +1,102 @@
+	return redskyv1beta1.Metric{
+		Type:     redskyv1beta1.MetricPrometheus,
+		Query:    query,
+		Minimize: true,
```
not something for this pr, but I imagine as we add in additional objectives ( ex throughput / requests/s ) we'll want to poke at this.

jgustie replied:
This just ends up being the starting point, the caller can always adjust it as necessary (e.g. the `DurationMetricsSource` overwrites the `Type` to be `""`).

bradbeam commented on internal/experiment/generation/generation.go:
```diff
@@ -0,0 +1,102 @@
+		},
+	}
+
+	if err := getter.GetFile(dst, src, opts...); err != nil {
```
kind of surprised go-getter doesn't support an io.Writer for the destination. oh well.

jgustie replied:
This was intensely frustrating to me. I was sure that it was just something I was missing. But as best I can tell "go getter" is really more like "go copier".

bradbeam commented on internal/experiment/generation/locust.go:
```diff
@@ -0,0 +1,182 @@
+	}
+
+	exp.Spec.TrialTemplate.Spec.JobTemplate = &batchv1beta1.JobTemplateSpec{}
+	pod := &exp.Spec.TrialTemplate.Spec.JobTemplate.Spec.Template.Spec
```
👀

bradbeam replied:
Should we add a method to Experiment to make this a bit more concise ( ex, Example.TrialPodSpec() *corev1.Pod )?

bradbeam commented on internal/experiment/generation/locust.go:
```diff
@@ -0,0 +1,182 @@
+
+		case obj.ErrorRate != nil:
+			if obj.ErrorRate.ErrorRateType == redskyappsv1alpha1.ErrorRateRequests {
+				query := `scalar(failure_count{job="trialRun",instance="{{ .Trial.Name }}"} / request_count{job="trialRun",instance="{{ .Trial.Name }}"})`
```
not a today thing, but I feel like it'd be worthwhile to split these out and group them up separately so we'd have a library of prom queries.

bradbeam commented on internal/experiment/generation/locust.go:
```diff
@@ -0,0 +1,182 @@
+		},
+	}
+
+	// TODO We need to rethink how ingress scanning works, this just preserves existing behavior
+	var ingressURL string
+	if s.Application != nil && s.Application.Ingress != nil {
+		ingressURL = s.Application.Ingress.URL
+	}
+	if ingressURL == "" {
+		return fmt.Errorf("ingress must be configured when using Locust scenarios")
+	}
+	pod.Containers[0].Env = append(pod.Containers[0].Env, corev1.EnvVar{Name: "HOST", Value: ingressURL})
```
is it worth moving this to `locustEnv()`?

jgustie replied:
I wanted to leave the ingress stuff separate for the time being, I'm not sure if this is something that will even end up in here or if there will be something like an `IngressSource` that eventually sets it. Same is true in the StormForger code.

bradbeam commented on internal/experiment/generation/stormforger.go:
```diff
@@ -0,0 +1,255 @@
+
+var _ ExperimentSource = &LocustSource{} // Update trial job
+var _ MetricSource = &LocustSource{}     // StormForger specific metrics
+var _ kio.Reader = &LocustSource{}       // ConfigMap for the test case file
```
Should these be `StormForgerSource` ?

jgustie replied:
🤦

bradbeam commented on internal/experiment/generation/transformer.go:
```diff
@@ -0,0 +1,264 @@
+	}
+}
+
+// parameterNamer returns a name generation function for parameters based on scan results.
```
how boring, where is my `whale-steals-armageddon` ? :)

jgustie committed:
Address code review comments
